control-plane/connect-inject/controllers/pod/pod.go

```diff
@@ -0,0 +1,361 @@
+		//}
+
+		// TODO(dans): delete proxyConfiguration
+		//if err := r.deleteProxyConfiguration(ctx, pod); err != nil {
+		//	errs = multierror.Append(errs, err)
+		//}
+
+		// TODO: clean up ACL Tokens
+
+		// TODO(dans): delete health status, since we don't have finalizers
+		//if err := r.deleteHealthStatus(ctx, pod); err != nil {
+		//	errs = multierror.Append(errs, err)
+		//}
+
+		return ctrl.Result{}, errs
+	} else if err != nil {
+		r.Log.Error(err, "failed to get Pod", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("retrieved", "name", pod.Name, "ns", pod.Namespace)
+
+	if hasBeenInjected(pod) {
+		if err := r.writeWorkload(ctx, pod); err != nil {
+			errs = multierror.Append(errs, err)
+		}
+
+		// TODO(dans): create proxyConfiguration
+
+		// TODO: create explicit upstreams
+		//if err := r.writeUpstreams(ctx, pod); err != nil {
+		//	errs = multierror.Append(errs, err)
+		//}
+
+		// TODO(dans): write health status
+		//if err := r.writeHealthStatus(ctx, pod); err != nil {
+		//	errs = multierror.Append(errs, err)
+		//}
+	}
+
+	return ctrl.Result{}, errs
+}
+
+func (r *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		Complete(r)
+}
+
+// hasBeenInjected checks the value of the status annotation and returns true if the Pod has been injected.
+func hasBeenInjected(pod corev1.Pod) bool {
+	if anno, ok := pod.Annotations[constants.KeyMeshInjectStatus]; ok && anno == constants.Injected {
+		return true
+	}
+	return false
+}
+
+func (r *Controller) deleteWorkload(ctx context.Context, pod types.NamespacedName) error {
+	req := &pbresource.DeleteRequest{
+		Id: getWorkloadID(pod.Name, r.getConsulNamespace(pod.Namespace), r.getPartition()),
+	}
+
+	_, err := r.ConsulResourceServiceClient.Delete(ctx, req)
+	return err
+}
+
+func (r *Controller) deleteHealthStatus(ctx context.Context, pod corev1.Pod) error {
+	return nil
+}
+
+func (r *Controller) writeWorkload(ctx context.Context, pod corev1.Pod) error {
+
+	// TODO: we should add some validation on the required fields here
+	// e.g. what if token automount is disabled and there is not SA. The API call
+	// will fail with no indication to the user other than controller logs
+	ports, workloadPorts := getWorkloadPorts(pod)
+
+	var node corev1.Node
+	// Ignore errors because we don't want failures to block running services.
+	_ = r.Client.Get(context.Background(), types.NamespacedName{Name: pod.Spec.NodeName, Namespace: pod.Namespace}, &node)
+	locality := parseLocality(node)
+
+	workload := &pbcatalog.Workload{
+		Addresses: []*pbcatalog.WorkloadAddress{
+			{Host: pod.Status.PodIP, Ports: ports},
+		},
+		Identity: pod.Spec.ServiceAccountName,
+		Locality: locality,
+		NodeName: common.ConsulNodeNameFromK8sNode(pod.Spec.NodeName),
+		Ports:    workloadPorts,
+	}
+
+	proto, err := anypb.New(workload)
+	if err != nil {
+		return fmt.Errorf("could not serialize workload: %w", err)
+	}
```

Comment:
IMHO, I'd rather have a log than a panic, otherwise, I agree the helper makes sense. I'll add a TODO in my PR to come back and refactor this when you merge it.